### ISSUE TITLE
Experiencing slow recompilation when manually building XLA

### ISSUE THREAD

**Issue by wenboqian** [pytorch/xla#7057]
## ❓ Questions and Help

Hi, I am interested in contributing to XLA community but I encounter a small challenge. After manually building `torch` and `torch_xla` on a CPU-based(CPU: **Intel(R) Xeon(R) Platinum 8375C CPU @ 2.90GHz**) Docker env, I noticed that the `python setup.py develop` process will take about **1 minutes** each time. So could you suggest any Dockerfile configurations or other changes that might speed up the recompilation process? Thanks for your help!

**Comment by JackCaoG** [comment]
if you didn't change any C++ files, you don't need to recompile. If you do modified the C++ files, depending on which file you change, 1 minutes recompilation times sounds about right. @will-cromar any thoughts?

**Comment by will-cromar** [comment]
Yeah, I agree. 1 minute is about typical for small C++ changes on my machine. Linking `_XLAC.so` is surprisingly slow.

The settings `XLA_CUDA=1`, `BUILD_CPP_TESTS=1`, and `BUNDLE_LIBTPU=1` can slow down build noticeably, so make sure those are unset or set to `0` in your image. But, if your rebuild is only taking 1 minute, they are probably already disabled.

**Comment by wfckl789** [comment]
By the way, any suggestion on how to print variables' value during runtime more efficiently? I am not so familiar with c++ so glad to hear input from your side :)

**Comment by bhavya01** [comment]
@wfckl789 Printing the values of torch xla tensors will always cause the execution of XLA graph. If you're just trying to develop a model, I think the performance at that stage won't be a concern but once you're convinced about the accuracy of your setup, then maybe reduce the number of print statements?

**Comment by wfckl789** [comment]
I see. Thanks for the advice! 

But I mean if I need to do some debugging in torch_xla `.cpp` files such as using `TF_VLOG()` to print C++ variables, could you please offer some suggested actions on how to observe variables' **real value in C++** during run time?  👀

**Comment by bhavya01** [comment]
I think that you will have to execute the graph to get the real value of the variables. Maybe this test can help https://github.com/pytorch/xla/blob/master/test/cpp/test_xla_backend_intf.cpp#L58

**Comment by wfckl789** [comment]
> I think that you will have to execute the graph to get the real value of the variables. Maybe this test can help https://github.com/pytorch/xla/blob/master/test/cpp/test_xla_backend_intf.cpp#L58

Thanks, this hint is very helpful to me! By the way, could you give some hints on which modules are responsible for "Python code tracing part" shown below? 🤔 
![image](https://github.com/pytorch/xla/assets/42508752/63a50af8-f5fc-471d-915f-b1dec00f90aa)

**Comment by bhavya01** [comment]
If you're using dynamo, then the dynamo_ module in Pytorch/pytorch repo does the python code tracing. It is helpful to go through an example using pdb to understand what is happening.

Otherwise, it's just the regular python interpreter.

**Comment by wfckl789** [comment]
> If you're using dynamo, then the dynamo_ module in Pytorch/pytorch repo does the python code tracing. It is helpful to go through an example using pdb to understand what is happening.
> 
> Otherwise, it's just the regular python interpreter.

I see. Could you tell me how to debug XLA's c++ code? I found using vscode debug extension can't trace c++ part.

**Comment by ysiraichi** [comment]
@wenboqian Do you still need any help on this issue?
